Assets/Plugins/iOS/Shopify/WebCheckoutViewController.swift

```diff
@@ -0,0 +1,236 @@
+        webView.layer.masksToBounds = true
+        webView.layer.cornerRadius = cornerRadius
+        
+        cancelButton = UIButton(type: .custom)
+        cancelButton.setTitleColor(.white, for: .normal)
+        cancelButton.titleLabel?.font = UIFont.boldSystemFont(ofSize: 18)
+        cancelButton.layer.cornerRadius = cornerRadius
+        cancelButton.translatesAutoresizingMaskIntoConstraints = false
+        
+        loadingView = LoadingView()
+        loadingView.translatesAutoresizingMaskIntoConstraints = false
+        loadingView.layer.cornerRadius = cornerRadius
+        
+        super.init(frame: frame)
+        
+        addSubview(webView)
+        addSubview(cancelButton)
+        addSubview(loadingView)
+        
+        NSLayoutConstraint.activate([
+            webView.topAnchor.constraint(equalTo: topAnchor, constant: 20),
+            webView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10),
+            webView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10),
+            
+            cancelButton.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10),
+            cancelButton.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -10),
+            cancelButton.topAnchor.constraint(equalTo: webView.bottomAnchor, constant: 10),
+            cancelButton.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -10),
+            
+            loadingView.topAnchor.constraint(equalTo: webView.topAnchor),
+            loadingView.bottomAnchor.constraint(equalTo: webView.bottomAnchor),
+            loadingView.trailingAnchor.constraint(equalTo: webView.trailingAnchor),
+            loadingView.leadingAnchor.constraint(equalTo: webView.leadingAnchor)
+        ])
+    }
+    
+    override func didMoveToWindow() {
+        super.didMoveToWindow()
+        setButton(state: .checkingOut)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    func setButton(state: CheckoutButtonState) {
+        assert(Thread.isMainThread, "Updating the button must be done on the main thread.")
+        switch state {
+        case .checkingOut:
+            self.cancelButton.backgroundColor = .red
+            self.cancelButton.setTitle("Cancel", for: .normal)
+        case .finished:
+            self.cancelButton.backgroundColor = .green
+            self.cancelButton.setTitle("Finished", for: .normal)
+        }
+    }
+    
+    func startInitialLoad() {
+        loadingView.alpha = 1
+        loadingView.indicator.startAnimating()
+        initialLoading = true
+    }
+    
+    func endInitialLoad() {
+        UIView.animate(withDuration: 0.3) {
+            self.loadingView.alpha = 0
+            self.initialLoading = false
+        }
+    }
+}
+
+// Loading view overlay presented on top of the web view when its first loading.
+private class LoadingView: UIView {
```

Comment:
I add this loading view on top of the web view on initial load to make the experience a little cleaner while the user is waiting for the page to render. I'm using the default loading indicator but we can do something custom here. We can also expose this to the Unity developer to add their own branded loading indicator.